Add "ignore leaf topics" option to recorder

ivanpauno opened the pull request:
Fixes https://github.com/ros2/rosbag2/issues/912.

emersonknapp commented on rosbag2_transport/src/rosbag2_transport/topic_filter.hpp:
```diff
@@ -56,6 +58,12 @@ filter_topics_with_known_type(
 
+ROSBAG2_TRANSPORT_PUBLIC
+std::unordered_map<std::string, std::string>
+filter_leaf_topics(
```
Can you add a docstring for this please? I know the other functions don't have them (but that's a bad and confusing thing)

ivanpauno replied:
Sounds good, fixed in https://github.com/ros2/rosbag2/pull/925/commits/da80c562101222cfb97583369de5811cb50e38be.

emersonknapp reviewed (approved):
This LGTM - although I wish it could be slightly less verbose to make one of those tests, that's not your problem to solve with this change.

ivanpauno commented:
CI:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=15691)](http://ci.ros2.org/job/ci_linux/15691/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=10371)](http://ci.ros2.org/job/ci_linux-aarch64/10371/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=13355)](http://ci.ros2.org/job/ci_osx/13355/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=15927)](http://ci.ros2.org/job/ci_windows/15927/)

ivanpauno commented:
I have fixed the deprecation warning that CI catched in https://github.com/ros2/rosbag2/pull/925/commits/c9d9333e0a06f8d0a8c4d432d888c451d2bed03e.

For some reason, the warning is only being emited in windows (some weird interaction between pragmas and templates, I've seen that before ...).

I have also rebased to resolve all the conflicts that were generated when https://github.com/ros2/rosbag2/pull/924 was merged.

ivanpauno commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=15698)](http://ci.ros2.org/job/ci_linux/15698/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=10378)](http://ci.ros2.org/job/ci_linux-aarch64/10378/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=13365)](http://ci.ros2.org/job/ci_osx/13365/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=15934)](http://ci.ros2.org/job/ci_windows/15934/)

emersonknapp commented:
Thanks for figuring out the merge - sorry about any difficulty that caused.
